File: aiops-contract/python/aiops_contract/report.py
"""
AIOps Report Contract — Pydantic Schema Definitions

共同語言：Agent 與 AIOps 之間的溝通標準。
"""

from __future__ import annotations

from typing import Dict, Any, List, Literal, Optional, Union
from pydantic import BaseModel, Field


SCHEMA_VERSION = "aiops-report/v1"


# ---------------------------------------------------------------------------
# Evidence Chain
# ---------------------------------------------------------------------------

class EvidenceItem(BaseModel):
    """單一工具執行結果對應的證據條目。"""

    step: int = Field(..., description="執行順序（從 1 開始）")
    tool: str = Field(..., description="mcp_name 或 skill_id")
    finding: str = Field(..., description="一句話結論，給人類閱讀")
    viz_ref: Optional[str] = Field(
        None, description="對應 visualization[].id，可 null"
    )


# ---------------------------------------------------------------------------
# Visualization
# ---------------------------------------------------------------------------

class VisualizationItem(BaseModel):
    """單一視覺化區塊，type 決定前端使用哪個 renderer。"""

    id: str = Field(..., description="唯一識別，供 evidence_chain.viz_ref 引用")
    type: str = Field(
        ...,
        description=(
            "renderer 類型。"
            "標準值：vega-lite | kpi-card | topology | gantt | table。"
            "未知 type 前端顯示 unsupported placeholder。"
        ),
    )
    spec: Dict[str, Any] = Field(
        ...,
        description="對應 type 的 spec。vega-lite 使用標準 Vega-Lite JSON spec。",
    )


# ---------------------------------------------------------------------------
# Suggested Actions
# ---------------------------------------------------------------------------

class AgentAction(BaseModel):
    """重新觸發 Agent 的動作。"""

    label: str
    trigger: Literal["agent"]
    message: str = Field(..., description="帶入 Agent 的 next message")


class HandoffAction(BaseModel):
    """移交給 AIOps 處理的動作，AIOps 接管後續 UI 互動。"""

    label: str
    trigger: Literal["aiops_handoff"]
    mcp: str = Field(..., description="AIOps Handoff MCP name")
    params: Optional[Dict[str, Any]] = None


SuggestedAction = Union[AgentAction, HandoffAction]


# ---------------------------------------------------------------------------
# Root Contract
# ---------------------------------------------------------------------------

class AIOpsReportContract(BaseModel):
    """
    AIOps Report Contract v1

    Agent 輸出的標準化結果，任何實作此 Contract 的前端皆可渲染。
    """

    schema_version: str = Field(
        default=SCHEMA_VERSION,
        alias="$schema",
        description="Contract 版本識別",
    )
    summary: str = Field(..., description="給人類閱讀的根因結論或回應摘要")
    evidence_chain: List[EvidenceItem] = Field(
        default_factory=list,
        description="推理過程中每個工具呼叫的關鍵發現",
    )
    visualization: List[VisualizationItem] = Field(
        default_factory=list,
        description="視覺化區塊列表",
    )
    suggested_actions: List[SuggestedAction] = Field(
        default_factory=list,
        description="建議的後續動作，前端渲染為可點擊按鈕",
    )

    model_config = {"populate_by_name": True}

File: aiops-contract/typescript/src/report.ts
/**
 * AIOps Report Contract — TypeScript Type Definitions
 *
 * 共同語言：Agent 與 AIOps 之間的溝通標準。
 */

export const SCHEMA_VERSION = "aiops-report/v1" as const;

// ---------------------------------------------------------------------------
// Evidence Chain
// ---------------------------------------------------------------------------

export interface EvidenceItem {
  /** 執行順序（從 1 開始） */
  step: number;
  /** mcp_name 或 skill_id */
  tool: string;
  /** 一句話結論，給人類閱讀 */
  finding: string;
  /** 對應 visualization[].id，可 undefined */
  viz_ref?: string;
}

// ---------------------------------------------------------------------------
// Visualization
// ---------------------------------------------------------------------------

/**
 * 標準 visualization type 值。
 * 前端未認識的 type 顯示 UnsupportedPlaceholder，不 crash。
 */
export type VisualizationType =
  | "vega-lite"
  | "kpi-card"
  | "topology"
  | "gantt"
  | "table"
  | (string & {}); // 允許擴充的自訂 type

export interface VisualizationItem {
  /** 唯一識別，供 evidence_chain.viz_ref 引用 */
  id: string;
  /** renderer 類型 */
  type: VisualizationType;
  /**
   * 對應 type 的 spec。
   * - vega-lite：標準 Vega-Lite JSON spec
   * - 其他：自訂 schema，前端對應 component 負責解析
   */
  spec: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Suggested Actions
// ---------------------------------------------------------------------------

export interface AgentAction {
  label: string;
  trigger: "agent";
  /** 帶入 Agent 的 next message */
  message: string;
}

export interface HandoffAction {
  label: string;
  trigger: "aiops_handoff";
  /** AIOps Handoff MCP name */
  mcp: string;
  params?: Record<string, unknown>;
}

export type SuggestedAction = AgentAction | HandoffAction;

// ---------------------------------------------------------------------------
// Root Contract
// ---------------------------------------------------------------------------

export interface AIOpsReportContract {
  $schema: typeof SCHEMA_VERSION;
  /** 給人類閱讀的根因結論或回應摘要 */
  summary: string;
  /** 推理過程中每個工具呼叫的關鍵發現 */
  evidence_chain: EvidenceItem[];
  /** 視覺化區塊列表 */
  visualization: VisualizationItem[];
  /** 建議的後續動作，前端渲染為可點擊按鈕 */
  suggested_actions: SuggestedAction[];
}

// ---------------------------------------------------------------------------
// Type Guards
// ---------------------------------------------------------------------------

export function isAgentAction(action: SuggestedAction): action is AgentAction {
  return action.trigger === "agent";
}

export function isHandoffAction(action: SuggestedAction): action is HandoffAction {
  return action.trigger === "aiops_handoff";
}

export function isValidContract(value: unknown): value is AIOpsReportContract {
  if (typeof value !== "object" || value === null) return false;
  const obj = value as Record<string, unknown>;
  return (
    obj["$schema"] === SCHEMA_VERSION &&
    typeof obj["summary"] === "string" &&
    Array.isArray(obj["evidence_chain"]) &&
    Array.isArray(obj["visualization"]) &&
    Array.isArray(obj["suggested_actions"])
  );
}
